discovery_engine_core/core/src/engine.rs

```diff
@@ -210,6 +223,124 @@ where
+                    .map_err(|err| Error::Ranker(err.into()))?,
+            )
+            .map_err(|err| Error::Ranker(err.into()))?
+            .with_accents(AccentChars::Cleanse)
+            .with_case(CaseChars::Lower)
+            .with_pooling::<AveragePooler>()
+            .build()
+            .map_err(GenericError::from)?;
+        let coi = de_config
+            .extract::<CoiSystemConfig>()
+            .map_err(|err| Error::Ranker(err.into()))?
+            .build();
+        let kpe = KpeConfig::from_files(
+            &config.kpe_vocab,
+            &config.kpe_model,
+            &config.kpe_cnn,
+            &config.kpe_classifier,
+        )
+        .map_err(|err| Error::Ranker(err.into()))?
+        .with_token_size(
+            de_config
+                .extract_inner("kpe.token_size")
+                .map_err(|err| Error::Ranker(err.into()))?,
+        )
+        .map_err(|err| Error::Ranker(err.into()))?
+        .with_accents(AccentChars::Cleanse)
+        .with_case(CaseChars::Keep)
+        .build()
+        .map_err(GenericError::from)?;
+        let providers = Providers::new(&config.clone().into()).map_err(Error::ProviderError)?;
+
+        // read the configs
+        let endpoint_config = de_config
+            .extract_inner::<EndpointConfig>("endpoint")
+            .map_err(|err| Error::Ranker(err.into()))?
+            .with_init_config(config)
+            .await;
+        let core_config = de_config
+            .extract_inner("core")
+            .map_err(|err| Error::Ranker(err.into()))?;
+        let exploration_config = de_config
+            .extract_inner(&format!("stacks.{}", Exploration::id()))
+            .map_err(|err| Error::Ranker(err.into()))?;
+
+        // set the states
+        let stack_ops = vec![
+            Box::new(BreakingNews::new(
+                &endpoint_config,
+                providers.headlines.clone(),
+            )) as _,
+            Box::new(TrustedNews::new(
+                &endpoint_config,
+                providers.trusted_headlines.clone(),
+            )) as _,
+            Box::new(PersonalizedNews::new(
+                &endpoint_config,
+                providers.news.clone(),
+            )) as _,
+        ];
+        let (mut stack_data, state) = if let Some(state) = state {
+            if stack_ops.is_empty() {
+                return Err(Error::NoStackOps);
+            }
+            SerializedState::deserialize(state)?
+        } else {
+            (HashMap::default(), CoiSystemState::default())
+        };
+        for id in stack_ops.iter().map(Ops::id).chain(once(Exploration::id())) {
+            if let Ok(alpha) = de_config.extract_inner::<f32>(&format!("stacks.{id}.alpha")) {
+                stack_data.entry(id).or_default().alpha = alpha;
+            }
+            if let Ok(beta) = de_config.extract_inner::<f32>(&format!("stacks.{id}.beta")) {
+                stack_data.entry(id).or_default().beta = beta;
+            }
```

Comment:
:thinking:  Isn't that wrong?

I mean it's outside of the scope of the PR and even if it is wrong shouldn't be fixed in this PR but shouldn't config `stacks.{id}.alpha` only be the default alpha value. Instead here it will always reset it to this value every time the engine is created overwriting whatever value was deserialized (on line 305).
 
I.e. shouldn't it be `.or_insert(alpha)` ?

Reply:
the current implementation mirrors the previous behavior (from before the config was changed in #466 and the constructors were refactored in #465), in `Engine::from_config()` we have that
- if no `state` is present, then a default stack data is created (as in the old `Engine::new()`)
- if some `state` is present, then this is taken (as in the old `Engine::from_state()`)
- if we have a value present in the `de_config`, then this is by design meant to overwrite any existing or default value, which happens in the marked lines. this design can of course be changed if needed.

Reply:
> if we have a value present in the de_config, then this is by design meant to overwrite any existing or default value, which happens in the marked lines. this design can of course be changed if needed.

I didn't know about this ;)